### PR TITLE
[Enhance] Better YOLOv3 head initialization

### PR DESCRIPTION
**Pull request by RangiLyu**
# Better YOLOv3 Head Initialization
## Motivation
The original YOLOv3 in mmdet initializes bias in convs_pred with 0, which results in a very large loss in the early training steps. And this will cause **overflow during fp16 training**.

## Modification
Use `bias_init_with_prob(0.01)` to initialize the bias of class and use a prior of "8 objects per feature map" to initialize objectness bias. Inspired by https://github.com/ultralytics/yolov3 but not the same.

## Effect
After initializing head with this method, the loss of classification and objectness reduced a lot at the early training steps.

Here are losses of epoch1 step50:
Before
```
loss_cls: 4499.7637, loss_conf: 94941.6641, loss_xy: 226.8941, loss_wh: 80.2181, loss: 45748.5430
```
After
```
loss_cls: 308.9337, loss_conf: 590.1021, loss_xy: 159.6315, loss_wh: 38.8394, loss: 1097.5066
```

|    Model     |  Use New Init   |  Type   | mAP | AP50| AP75| AP s| AP m | AP l |
:--------------: | :------: | :------: | :------------: | :----: | :-----: | :------: | :--------: | :--------: |
yolov3_d53_mstrain-608_273e_coco| No | FP32 | 33.4 | 56.3 | 35.2 | 19.5 | 36.4 | 43.6 |
yolov3_d53_mstrain-608_273e_coco| No | FP16(Overflow) | - | - | - | - | - | - |
yolov3_d53_mstrain-608_273e_coco| Yes | FP32 | 33.7(+0.3) | 56.6(+0.3) | 35.3(+0.1) | 19.4 | 36.8(+0.4) | 44.3(+0.7) |
yolov3_d53_mstrain-608_273e_coco| Yes | FP16 |33.8(+0.4) | 56.8(+0.5) | 35.1| 19.7(+0.2) | 37.3(+0.9) | 43.6 |

### CODE REVIEW

File: mmdet/models/dense_heads/yolo_head.py [hunk 1]
```diff
@@ -147,6 +148,22 @@ def _init_layers(self):
...
+    def init_weights(self):
+        for m in self.modules():
+            if isinstance(m, nn.Conv2d):
+                normal_init(m, mean=0, std=0.01)
+            if is_norm(m):
+                constant_init(m, 1)
+
+        # Use prior in model initialization to improve stability
+        for conv_pred, stride in zip(self.convs_pred, self.featmap_strides):
+            bias = conv_pred.bias.reshape(self.num_anchors, -1)
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/yolo_head.py:160]: Is this line necessary?

**Comment by RangiLyu** [comment]
## Update

|    Model     |  Use New Init   |  Type   | mAP | AP50| AP75| AP s| AP m | AP l |
:--------------: | :------: | :------: | :------------: | :----: | :-----: | :------: | :--------: | :--------: |
yolov3_d53_mstrain-608_273e_coco| No | FP32 | 33.4 | 56.3 | 35.2 | 19.5 | 36.4 | 43.6 |
yolov3_d53_mstrain-608_273e_coco| No | FP16(Overflow) | - | - | - | - | - | - |
yolov3_d53_mstrain-608_273e_coco| Yes | FP32 | 33.7(+0.3) | 56.6(+0.3) | 35.3(+0.1) | 19.4 | 36.8(+0.4) | 44.3(+0.7) |
yolov3_d53_mstrain-608_273e_coco| Yes | FP16 |33.8(+0.4) | 56.8(+0.5) | 35.1| 19.7(+0.2) | 37.3(+0.9) | 43.6 |

It seems that the results will be a little bit higher after using this initialization method.

File: mmdet/models/dense_heads/yolo_head.py (hunk 1 above)
Comment (RangiLyu) [reply mmdet/models/dense_heads/yolo_head.py:160]: The origin shape of the bias is [255]. It would be easier to initialize each anchor's classification bias and objectness bias after reshape to [3, 85]. Otherwise, the index will be complicated.

**Review by ZwwWayne — APPROVED** [review]

**Comment by ZwwWayne** [comment]
Can be merged after fix unit tests.

**Comment by RangiLyu** [comment]
> Can be merged after fix unit tests.

Unit tests failed during testing onnx yolov3 neck, but I didn't change that part. I just re-run and passed. Seems to be a random bug.
